src/bin/psql/model.c

```diff
@@ -0,0 +1,174 @@
+#include "model.h"
+
+#include <stdio.h>
+#include <string.h>
+#include "adhoc_compat.h"
+
+void ModelUpdateLens(Model *m, Row* r)
+{
+	size_t i = 0;
+
+	if (m->nfields < RowSize(r))
+	{
+		m->maxlens = realloc(m->maxlens, RowSize(r) * sizeof(size_t));
+
+		for (i = m->nfields; i < RowSize(r); ++i)
+		{
+			m->maxlens[i] = 0;
+		}
+
+		m->nfields = RowSize(r);
+	}
+
+	for (i = 0; i < m->nfields; ++i)
+	{
+		m->maxlens[i] = MAX(m->maxlens[i], RowFieldLength(r, i)); 
+	}
+}
+
+Model* ModelInit()
+{
+	Model* m = malloc(sizeof(Model));
+	memset(m, 0, sizeof(Model));
+
+	m->rowmap = RowMapInit();
+
+	return m;
+}
+
+void ModelDestroy(Model *m)
+{
+	RowMapDestroy(m->rowmap);
+	RowCleanup(&m->header);
+
+	free(m->maxlens);
+	memset(m, 0, sizeof(Model));
+
+	free(m);
+
+	RowKeyReset();
+}
+
+void ModelAddRow(Model *m, Row* r)
+{
+	ModelUpdateLens(m, r);
+	RowMapUpdate(m->rowmap, r);
+}
+
+void ModelDeleteRow(Model *m, Row* r)
+{
+	RowMapErase(m->rowmap, r);
+	RowCleanup(r);
+}
+
+static size_t FindInRow(Row *r, const char* s)
+{
+	size_t i = 0;
+
+	for (i = 0; i < RowSize(r); ++i)
+	{
+		if (strcmp(RowFieldValue(r,i), s) == 0)
+		{
```

Comment:
Per the style guide, brackets aren't necessary for single statements under control structures:

```
if (strcmp(RowFieldValue(r,i), s) == 0)
    break;
```

http://www.postgresql.org/docs/current/static/source.html